mlir/lib/Dialect/Bufferization/IR/BufferizationDialect.cpp

```diff
@@ -63,6 +64,9 @@ void mlir::bufferization::BufferizationDialect::initialize() {
 #include "mlir/Dialect/Bufferization/IR/BufferizationOps.cpp.inc"
       >();
   addInterfaces<BufferizationInlinerInterface>();
+  declarePromisedInterface<func::CallOp, BufferizableOpInterface>();
+  declarePromisedInterface<func::FuncOp, BufferizableOpInterface>();
+  declarePromisedInterface<func::ReturnOp, BufferizableOpInterface>();
```

Comment:
Indeed. @justinfargnoli was it intentional to declare these here for `func`, but not for `arith`, for example?

The general case hits the crux of the issue here, which is "who is allowed to attach (or promise) interfaces for an op". Intuitively, it should be either the "owner" of the op or the "owner" of the interface, but I can see cases where it would be neither of those.

Reply:
Yes, it was intentional. I was following the format of `registerBufferizableOpInterfaceExternalModels()` - called [here](https://github.com/llvm/llvm-project/blob/7dd790db8b77c4a833c06632e903dc4f13877a64/mlir/include/mlir/InitAllDialects.h#L150) and defined [here](https://github.com/llvm/llvm-project/blob/7dd790db8b77c4a833c06632e903dc4f13877a64/mlir/lib/Dialect/Bufferization/Transforms/FuncBufferizableOpInterfaceImpl.cpp#L490). 

My understanding was that this was a dialect extension for the `Bufferization` dialect. Thus, they should be declared in the `Bufferization` dialect's initialization. However, I don't have a good understanding of how dialect extensions work, so I'm likely wrong. (Links to documentation would be greatly appreciated!)

Reply:
Extensions are basically callbacks that trigger when the corresponding dialect(s) are loaded. The only documentation we have for those is doxygen, I suppose.

Reply:
Pinging @ftynse and @joker-eph to determine if this thread can be marked as resolved or if we want to take a different approach.

Reply:
No I think we should be uniform here: the FuncDialect isn't more special than any other, and you defined the promise for all dialect in their own initializer. This is the only exception right now.